### PR TITLE
Modified Modules UI revamp with other UI changes

### PR DESCRIPTION
**Pull request by tgqiang**
Branch: **sprintFiveGq**

What was done in this branch (*not in chronological order*):
1. Massive revamp to Modified Modules UI
2. 404 Error Page minor restyling
3. Addressed #38 with https://github.com/nus-mtp/cs-modify/commit/2d08b205e7b9e74cc9d4fc26ee95b594557af551
4. Addressed #51 with https://github.com/nus-mtp/cs-modify/commit/afeb9faa1df06d381e3f7a3f022fe4cf0d625bca
5. Addressed #41 with https://github.com/nus-mtp/cs-modify/commit/5a98d4c066349589dd966e6298e8ff782c54b17d
6. Adjusted column widths for some DataTables through `csmodify.js`

### CODE REVIEW

File: test/test_add_module.py [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 import web
```
Comment (helloqx) [inline test/test_add_module.py]: Unused import web

File: test/test_add_module.py [hunk 2]
```diff
@@ -87,3 +115,170 @@ def test_add_module_page_content(self):
...
+
+        addModuleForm = root.forms__get()['addModForm']
+        addModuleForm.__setitem__('code', self.TEST_ADD_MODULE_CODE_RELAXED)
+        addModuleForm.__setitem__('name', self.TEST_ADD_MODULE_NAME)
+        addModuleForm.__setitem__('description', self.TEST_ADD_MODULE_DESCRIPTION)
+        addModuleForm.__setitem__('mc', self.TEST_ADD_MODULE_MC)
+
+        response = addModuleForm.submit()
+
+        response.mustcontain((self.OUTCOME_SUCCESS_MESSAGE % self.TEST_ADD_MODULE_CODE_RELAXED.upper()))
```
Comment (helloqx) [inline test/test_add_module.py]: 157, 158: line too long

File: test/test_edit_module_fixture.py [hunk 3]
```diff
@@ -79,26 +134,29 @@ def test_module_edit_submit_response(self):
...
-        root = self.test_app.get(self.URL_INDIVIDUAL_MODULE_VIEW)
-        assert_equal(root.status, 200)
-        submit_button = root.forms__get()[self.EDIT_MODULE_BUTTON_FORM_NAME]
-        response = submit_button.submit()
+        root = self.test_app.get(self.URL_EDIT_MODULE_SPECIFIC_VALID)
+        edit_module_mounting_form = root.forms__get()["edit-mounting-form"]
+        edit_module_mounting_form.__setitem__("quota", -1)
+        response = edit_module_mounting_form.submit()
 
+        print(response)
```
Comment (helloqx) [inline test/test_edit_module_fixture.py]: Remove debug statement?

File: test/test_module_modified.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        '''
+            Tests if viewing page displaying invalid module
+            changes for a valid module should fail.
+
+            Expected view is an empty table.
+        '''
+        root = self.test_app.get(self.URL_INVALID_CHANGE_TYPE)
+        assert_equal(root.status, 200)
+        root.mustcontain("Modified Mounting of <b>MM1009</b>")
+        root.mustcontain("<tr><td colspan=4>No data found</td></tr>")
```
Comment (helloqx) [inline test/test_module_modified.py]: Final newline missing

File: test/test_view_individual_module.py [hunk 5]
```diff
@@ -128,68 +127,18 @@ def test_view_individual_module_invalid_code_response(self):
...
-    # @raises(Exception)
-    # def test_view_invalid_module_overview_invalid_quota_response(self):
-    #     # Quota used here is '70' (actual is '60')
-    #     # an exception WILL be encountered here
-    #     root = self.testApp.get(self.URL_CONTAIN_CODE_AY_INVALID_QUOTA)
-    # '''
+    '''
+        Tests if user will fail to access page for showing module overview
+        if the target AY-semester is invalid.
+    '''
```
Comment (helloqx) [inline test/test_view_individual_module.py]: Shift docstring below def statement for consistency

File: test/test_view_module.py [hunk 6]
```diff
@@ -143,42 +145,36 @@ def test_view_module_overview_goto_individual_module_invalid_code(self):
...
-    #     response = root.goto(url, method='get')
-    # '''
+        assert_equal(response.status, 200)
+        response.mustcontain("Not Found")
 
 
-    # '''
-    #     Tests if navigation to an individual module view
-    #     with invalid quota is unsuccesful.
+    '''
```
Comment (helloqx) [inline test/test_view_module.py]: 154, 170: Shift docstring below def for consistency

**Comment by tgqiang** [comment]
@helloqx 
> Non-essential:
> The regex is very harsh, and the help message 'Please match the requested format' is too general

Can list for which pages particularly?

**Comment by helloqx** [comment]
Module Code field in Add Module

**Comment by xaterz** [comment]
image
Not sure if the cross should be underlined when hovering cursor over it.

**Comment by xaterz** [comment]
Issue #41 'capitalize all module codes for user' has been resolved for prior module and add module. However other pages should also be patched to ignore upper/lower cases too. For example in module overview, ?code=CS1010 works but ?code=cs1010 returns 404. Both should be acceptable.

Suggest to reopen issue. To be addressed in the future, not in this sprint/PR.

**Comment by tgqiang** [comment]
@xaterz @helloqx For the regex issue and issue #41, please raise them in issue tracker if required.

**Comment by xaterz** [comment]
'Modified Module Details for a target module' currently does not handle the case where the module is valid but does not have modified details. Please fix this bug

**Comment by tgqiang** [comment]
Bug fixed with new commits.

**Comment by helloqx** [comment]
Bugs have been fixed 👍
I feel that the > icon popping out the words can be a bit distracting when the user is browsing the sidebar. 
I think that the underline was less distracting.

**Comment by helloqx** [comment]
Needs #60

**Comment by tgqiang** [comment]
#60 addressed here with https://github.com/nus-mtp/cs-modify/pull/55/commits/786da5eae87ab7ac24d1c25a96e2173eeec86078.

**Pull request merged**